CHANGELOG.md

```diff
@@ -14,6 +14,9 @@ and what APIs have changed, if applicable.
 
 ## [Unreleased]
 
+## [29.20.0] - 2021-07-27
+- Fixed race condition when switching d2 load balancer strategies.
```

Comment:
So in the worst case, there will only be one request failure to trigger ring update, and all the subsequent requests  should succeed after the ring is updated. 

I was thinking to further eliminate the one-off failure, if `castToDegraderTrackerClients` fails, can we simply return any TrackerClient passed to the degrader strategy's `getTrackerClient`? It's similar to how ClientSelector works:

```
      // Pick one from the tracker clients passed from the request if the ring is completely out of date
      trackerClient = trackerClients.values().stream()
          .filter(latestTrackerClient -> !excludedUris.contains(latestTrackerClient.getUri()))
          .findAny().orElse(null);
```

Reply:
Basically i think we're missing this logic in Degrader strategy https://github.com/linkedin/rest.li/blob/7372bc76f7ce78007ecb24ca80f17399e5246e13/d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/ClientSelector.java#L137

Reply:
I think this is a very good idea to at least return something, but there is some logic that requires DegraderTrackerClient.

After a trackerClient is picked, there is a logic that checks whether the whole cluster is in the drop state, they requires the degraderTrackerClient instance to finish that check. It means if we pick a TrackerClientImpl instance, it will totally just skip that check, which will change the DegraderLoadBalancerV3 logic.

Now the trade-off becomes - should we try best effort to return a TrackerClientImpl instance without taking care of checkDrop() logic, or should we just let the request fail. What do you think?

Reply:
IMO, it would be better to at least return something other than fail a request. Given it's a one-off scenario, skipping checkDrop() seems okay to me. What are your thoughts?

Reply:
I'm not too worried about failing very few requests, but adding the check will make the casting logic more complicated. If it's not DegraderTrackerClient then we're using TrackerClientImpl again, so it uses different instances here and there, which makes the code a bit hard to follow. I don't have a strong opinion on this though, mainly because it will solve a very intermittent error.